### ISSUE TITLE
`test/checkpoint.bats` tests don't work on `arm64`

### ISSUE THREAD

**Issue by saschagrunert** [cri-o/cri-o#7629]
https://github.com/cri-o/cri-o/pull/7620 shows that the tests have to be disabled to let the CI pass on the `linux/arm64` platform. 

Goal of this issue is to investigate why the tests fail and fix them appropriately.

/help
/good-first-issue

**Comment by prakrit55** [comment]
@saschagrunert, I would like to give it a try

**Comment by prakrit55** [comment]
hey @saschagrunert, is still there a requirement  of the research and the issue

**Comment by prakrit55** [comment]
hey @saschagrunert, some files are like 
![Screenshot (239)](https://github.com/cri-o/cri-o/assets/98270250/c04b29f1-3c6b-4045-ba40-5a63fcbf2cb1)

and some tests couldn't identify these functions, is this a usual behaviour ?

**Comment by saschagrunert** [comment]
@prakrit55 no, but this issue talks about the integration tests, which use shell scripts to run in our CI: https://github.com/cri-o/cri-o/blob/main/test/checkpoint.bats

The issue is that we have to exclude them on our arm64 machines because they only seem to run on x86: https://github.com/cri-o/cri-o/blob/e9febd30c2e87024ac57fd59e01b2fc97e79f16d/test/checkpoint.bats#L10-L12

Maybe @adrianreber has a guess why this could be the case.

**Comment by adrianreber** [comment]
What kind of arm64 systems do you have. There is nothing preventing checkpoint/restore to work on arm64, but if the arm64 systems is maybe a container with limited access to the kernel some things might be blocked. Not sure.

**Comment by saschagrunert** [comment]
@adrianreber we're running on actuated: https://actuated.dev/blog/arm-ci-cncf-ampere

Your first guess could be right, there may be some limitation wrt host isolation.

**Comment by adrianreber** [comment]
Thanks for the link. It sounds like firecracker VMs, maybe they have a custom kernel which does not enable the necessary checkpoint restore interfaces. Hard to tell. They are not sharing many details.

**Comment by saschagrunert** [comment]
@adrianreber I have a test run here: https://github.com/cri-o/cri-o/actions/runs/7797927341/job/21265690481

Is this something we can work with:

```
msg="Response error: failed to checkpoint container
954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b: running
\"/usr/sbin/runc\" [\"checkpoint\" \"--file-locks\" \"--image-path\"
\"/tmp/tmp.KBrlVCRETn/crio/overlay-containers/954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b/userdata/checkpoint\"
\"--work-path\"
\"/tmp/tmp.KBrlVCRETn/crio/overlay-containers/954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b/userdata\"
\"--leave-running\"
\"954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b\"] failed:
`/usr/sbin/runc --root /tmp/tmp.KBrlVCRETn/crio-runtime-root --systemd-cgroup
checkpoint --file-locks --image-path
/tmp/tmp.KBrlVCRETn/crio/overlay-containers/954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b/userdata/checkpoint
--work-path
/tmp/tmp.KBrlVCRETn/crio/overlay-containers/954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b/userdata
--leave-running
954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b` failed:
time=\"2024-02-06T10:12:47Z\" level=error msg=\"criu failed: type NOTIFY errno
0\\nlog file:
/tmp/tmp.KBrlVCRETn/crio/overlay-containers/954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b/userdata/dump.log\"\n
: exit status 1" file="otel-collector/interceptors.go:71"
id=8de15d29-442b-41d7-b27c-415ada880583
name=/runtime.v1.RuntimeService/CheckpointContainer
```

**Comment by adrianreber** [comment]
Unfortunately that gives not enough information. That file would be interesting: `/tmp/tmp.KBrlVCRETn/crio/overlay-containers/954b8c78245381fb73e45f0c409bec85ad21050addd2941a16deb6d4a92eb81b/userdata/dump.log`.

The important information would be the kernel version and the kernel config and the content of that file.

If that information could be retrieved during the test I could say more. I am pretty sure the kernel has just some important bits disabled. So if we figure out what is missing the hard part would probably be still to convince the people behind these arm64 VMs to enable it.

**Comment by saschagrunert** [comment]
@adrianreber here we go: https://github.com/cri-o/cri-o/actions/runs/7799141677/job/21269407871?pr=7741

**Comment by prakrit55** [comment]
hey @saschagrunert, again this might be the reason too 
https://github.com/checkpoint-restore/criu/blob/495081c20ba23dc82bc83684730d36b5a02fdce5/criu/sockets.c#L210

and found this one too 
`# (00.000084) File /run/criu.kdat does not exist`, and couldn't get the file to sneak in

**Comment by adrianreber** [comment]
> `Error (compel/src/lib/ptrace.c:27): suspending seccomp failed: Invalid argument`

Yes, that is the problem. Looking at

https://github.com/torvalds/linux/blob/master/kernel/ptrace.c#L363

I am pretty sure they do not have `CONFIG_CHECKPOINT_RESTORE=y` in their kernel config.

Do you have any way to request that?

> `# (00.000084) File /run/criu.kdat does not exist, and couldn't get the file to sneak in`

@prakrit55 that is irrelevant

**Comment by saschagrunert** [comment]
Yep, I'm going to request the kernel flag in the actuated slack.

**Comment by saschagrunert** [comment]
@adrianreber we enabled the flag and it looks like the tests are now failing for a different reason: https://github.com/cri-o/cri-o/actions/runs/7799141677/job/21354250925?pr=7741

**Comment by adrianreber** [comment]
That was a quick turnaround, nice.

To fix the next errors something like this might be needed:

CONFIG_UNIX_DIAG=y
CONFIG_INET_DIAG=y
CONFIG_INET_UDP_DIAG=y
CONFIG_PACKET_DIAG=y
CONFIG_NETLINK_DIAG=y
CONFIG_NETFILTER_XT_MARK=y
CONFIG_INET_TCP_DIAG=y
CONFIG_INET_UDP_DIAG=y
CONFIG_INET_RAW_DIAG=y

There is an outdated page what needs to exist: https://criu.org/Linux_kernel

From that page I know that `CONFIG_MEM_SOFT_DIRTY=y` does not exist on arm64. Besides that most of those options should be turned on.

**Comment by adrianreber** [comment]
> @adrianreber is there anything else we need? @alexellis supports us with updating the machine 🙏

The things at https://criu.org/Linux_kernel should be all that is needed. That page is slightly outdated, so I am not 100% sure, I just know it works on all other major distribution kernels.

**Comment by saschagrunert** [comment]
@adrianreber it still seems to fail but the error message now looks different again. May I ask you to give those logs another look:

https://github.com/cri-o/cri-o/actions/runs/7927748050/job/21644773715?pr=7741

**Comment by alexellis** [comment]
Those settings were added as requested.

**Comment by adrianreber** [comment]
```
# (00.183859) Error (criu/cr-dump.c:203): 4063 has rseq but kernel lacks get_rseq_conf feature
```

Not sure which userspace is running on the kernel, but the CRIU binary in use is built with support for restartable sequences (and it has to). It seems the kernel is missing support to read the state of restartable sequences. That feature was merged later.  For the RHEL 8 kernel (4.18) it was an easy backport.

The whole story is that glibc enabled restartable sequence support once the kernel had `ptrace_get_rseq_configuration()`. To use CRIU on a system with a glibc enabled support for restartable sequence at least 5.13 kernel is necessary https://github.com/torvalds/linux/commit/90f093fa8ea48e5d991332cee160b761423d55c1.

From what I have seen the kernel we are talking about is a 5.10 kernel which is too old for  `ptrace_get_rseq_configuration()`. A backport should not be too difficult. The main problem is that the kernel is too old for the userspace components in use. It sounds like the given setup (5.10 kernel and 22.04 userspace is just a broken combination).

Possible workarounds. Tell glibc to disable support for restartable sequences: `GLIBC_TUNABLES=glibc.pthread.rseq=0`. Although I am not sure that will really work, because it needs to be set in the container CRI-O is trying to checkpoint/restore. Which means changing every container of every project testing checkpoint/restore. Another try could be to set `CONFIG_RSEQ=n` which means glibc will probably detect that it is not available and not use it.

So I would suggest either
* provide ` ptrace_get_rseq_configuration()`
or
* disable restartable sequence support

Adding @fweimer @mihalicyn in case for additional input.

@fweimer, would `CONFIG_RSEQ=n` help?

**Comment by adrianreber** [comment]
Another idea. @saschagrunert can you try following patch:
```diff
diff --git a/test/testdata/container_config.json b/test/testdata/container_config.json
index d3ba20d8b..daee7d114 100644
--- a/test/testdata/container_config.json
+++ b/test/testdata/container_config.json
@@ -20,6 +20,10 @@
                        "key": "TERM",
                        "value": "xterm"
                },
+               {
+                       "key": "GLIBC_TUNABLES",
+                       "value": "glibc.pthread.rseq=0"
+               },
                {
                        "key": "TESTDIR",
                        "value": "test/dir1"
diff --git a/test/testdata/container_config_ping.json b/test/testdata/container_config_ping.json
index 0530b735a..ff0e1d45d 100644
--- a/test/testdata/container_config_ping.json
+++ b/test/testdata/container_config_ping.json
@@ -16,6 +16,10 @@
                        "key": "PATH",
                        "value": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
                },
+               {
+                       "key": "GLIBC_TUNABLES",
+                       "value": "glibc.pthread.rseq=0"
+               },
                {
                        "key": "TERM",
                        "value": "xterm"
diff --git a/test/testdata/container_redis.json b/test/testdata/container_redis.json
index 8151ecfb4..a06dd0629 100644
--- a/test/testdata/container_redis.json
+++ b/test/testdata/container_redis.json
@@ -15,6 +15,10 @@
                        "key": "PATH",
                        "value": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
                },
+               {
+                       "key": "GLIBC_TUNABLES",
+                       "value": "glibc.pthread.rseq=0"
+               },
                {
                        "key": "TERM",
                        "value": "xterm"
diff --git a/test/testdata/container_sleep.json b/test/testdata/container_sleep.json
index 7c3a170cb..f42463183 100644
--- a/test/testdata/container_sleep.json
+++ b/test/testdata/container_sleep.json
@@ -16,6 +16,10 @@
                {
                        "key": "PATH",
                        "value": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
+               },
+               {
+                       "key": "GLIBC_TUNABLES",
+                       "value": "glibc.pthread.rseq=0"
                }
        ],
        "annotations": {
```

All tests in CRI-O are using those four test containers, so it is not as bad as I thought to adapt those containers to tell glibc to disable restartable sequences.

**Comment by saschagrunert** [comment]
@adrianreber yes this works, see: https://github.com/cri-o/cri-o/actions/runs/7957566854/job/21720736419

Do you think that this is a workaround we can use for now?

**Comment by adrianreber** [comment]
Cool. From my point of view that should be an acceptable workaround for now. Disabling "restartable sequences" from my understanding should not be a problem.

**Comment by saschagrunert** [comment]
Awesome, thank you so much @adrianreber :pray:

**Comment by mihalicyn** [comment]
> It should help. I think there was also a version of criu that could glean the RSEQ parameters from the glibc `__rseq_*` symbols, rather than going through the `ptrace` route?

We have this logic, but it's about `criu restore` stage, when we need to disable rseq for a current criu process to just before we restore the memory content.

But we still need to have this ptrace thing during `criu dump` to get an information about where `struct rseq` lives for all the processes, because we have no guarantee that `struct rseq` is placed in the same virtual address. And also I'm not sure how we can use this glibc `__rseq_*` symbols for non-current process. If you have any suggestions it would be awesome and I'll implement this.

**Comment by alexellis** [comment]
Hi @mihalicyn @fweimer @saschagrunert 

We have a 6.1 Kernel available for 64-bit Arm, when would you like to revisit this and test again?

https://docs.actuated.dev/faq/#what-kernel-version-is-being-used

**Comment by alexellis** [comment]
Related, the runc project is saying that criu is working for them on the older Kernel - https://github.com/opencontainers/runc/pull/4252

**Comment by adrianreber** [comment]
> Related, the runc project is saying that criu is working for them on the older Kernel - [opencontainers/runc#4252](https://github.com/opencontainers/runc/pull/4252)

As mentioned there, that is because their glibc is too old to user restartable sequences.
